### DISCUSSION TITLE
What is null-safety in dart programming

### DISCUSSION THREAD

**Discussion by erfanzar** [community/community#66578] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

describe null-safety in dart programming language and how does it works?
if you can please share an example for that please

**Comment by ehsanAghahoseini** [comment]
Sure. Null safety is a feature of the Dart programming language that prevents null pointer exceptions. A null pointer exception is an error that occurs when a variable is dereferenced (accessed) but contains the value null.

In Dart, all types are nullable by default. This means that a variable of any type can be null. However, you can explicitly mark a variable as non-nullable by adding a question mark (?) after the type name. For example, the following code declares a variable of type String?, which means that the variable can be null:

String? name;


If you try to dereference a variable that is null, the Dart compiler will generate an error. This prevents null pointer exceptions from occurring.

Here is an example of a null pointer exception in Dart:


String name;

print(name.toUpperCase()); // This will cause a null pointer exception because the variable name is null.


To fix this error, we can explicitly mark the variable name as non-nullable:


String? name;

name = "John Doe";

print(name.toUpperCase()); // This will now print "JOHN DOE" because the variable name is not null.


Null safety also introduces a number of new operators that can be used to handle null values. These operators include:

* The null-coalescing operator (??): This operator returns the value of the expression on the left if it is not null, or the value of the expression on the right if the expression on the left is null. For example, the following code assigns the value of the variable name to the variable greeting, if name is not null. Otherwise, it assigns the value "Hello, world!" to greeting:

String greeting = name ?? "Hello, world!";


* The safe navigation operator (?.): This operator returns the value of the expression on the left if it is not null, or null if the expression on the left is null. For example, the following code prints the value of the variable name if it is not null. Otherwise, it prints nothing:

print(name?.toUpperCase());


* The late keyword: This keyword can be used to declare a variable that will be initialized later. The variable must be non-nullable. For example, the following code declares a variable of type String called name. The variable will be initialized with the value of the user's input later.

**Comment by Akash1134** [comment]
GitHub's [Acceptable Use Policies](https://docs.github.com/en/site-policy/acceptable-use-policies/github-acceptable-use-policies) prohibits coordinated or inauthentic activity like rapid questions and answers. As a result, we'll be unmarking the answer and locking this post. Please note any future violations may result in a temporary or indefinite block from the Community. Thanks for understanding.
